### ISSUE TITLE
spray-json support unmarshaller charset problem

### ISSUE THREAD

**Issue by fehmicansaglam** [akka/akka#18389]
Hi,

I have experienced a strange problem and it seems to be a bug. Below is the request json that I send to a service:

```json
{"title":"ÇatÇat Kitapları","parentName":"Çocuk Kitapları"}
```

This json is decoded as:

```json
{"title":"Ã�atÇat Kitapları","parentName":"Ã�ocuk Kitapları"}
```

But more interestingly when I put a character before `Ç` as below:

```json
{"title":"aÇatÇat Kitapları","parentName":"aÇocuk Kitapları"}
```

It is decoded perfectly as below:

```json
{"title":"aÇatÇat Kitapları","parentName":"aÇocuk Kitapları"}
```

As I saw some very strange things are happening here I tried to debug the problem. Below is the code snippet that seems to be responsible for JsonParser creation in SprayJsonSupport.scala

```scala
Unmarshaller.byteStringUnmarshaller.forContentTypes(`application/json`).mapWithCharset { (data, charset) ⇒
  val input =
    if (charset == HttpCharsets.`UTF-8`) ParserInput(data.toArray)
    else ParserInput(data.decodeString(charset.nioCharset.name)) // FIXME: identify charset by instance, not by name!
  JsonParser(input)
}
```

So I tried changing the related line to this:

```scala
 if (charset == HttpCharsets.`UTF-8`) ParserInput(data.utf8String)
```

And everything worked perfectly. So am I doing smth wrong or is there a problem in spray-json or maybe in ByteString?

**Comment by ktoso** [comment]
Thanks for reporting @fehmicansaglam!
Sounds reasonable to me, however would like to ask @jrudolph or @sirthias for input on this.

**Comment by jrudolph** [comment]
spray-json can parse both from `String` and from `Array[Byte]` (which should be more efficient). Maybe parsing from `Array[Byte]` has a bug in which case parsing from `String` would be at least a valid workaround.

**Comment by ktoso** [comment]
Sounds like we have a plan. Would you like to submit the fix you mentioned @fehmicansaglam? 
It lives on the `release-2.3-dev` branch, thanks a lot in advance!

**Comment by fehmicansaglam** [comment]
@ktoso sure.

**Comment by ktoso** [comment]
Thanks @fehmicansaglam!

**Comment by sirthias** [comment]
@fehmicansaglam One question: Somehow I cannot reproduce the parsing problem you reported here.
What spray-json version were you using?

**Comment by fehmicansaglam** [comment]
@sirthias I could reproduce the problem in the test code here: https://github.com/akka/akka/pull/18390/files#diff-0c1eea43ece8e9f58a77044205b05124R22 
I can see that the version of spray-json is 1.3.1 here: https://github.com/fehmicansaglam/akka/blob/0891cd2bc4e8ac7d6bb371a14fa5f383404923e5/project%2FAkkaBuild.scala#L1551

**Comment by jrudolph** [comment]
Then it probably was https://github.com/spray/spray-json/issues/148 which was fixed in 1.3.2.

The dependency in akka was upgraded in https://github.com/akka/akka/commit/1db15763534614de48a13ea5011d87ae86f33afd. So, we should probably revert your original fix to the more efficient previous incantation which should work with spray-json 1.3.2.

**Comment by sirthias** [comment]
Yes, @jrudolph, I agree.
I'll take care of it and close all the tickets that have originated from this problem.
Thanks, @fehmicansaglam, for the clarification!
